client/app/bundles/course/assessment/submission/components/answers/Programming/ProgrammingFile.jsx

```diff
@@ -0,0 +1,100 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import { yellow100 } from 'material-ui/styles/colors';
+import Paper from 'material-ui/Paper';
+import WarningIcon from 'material-ui/svg-icons/alert/warning';
+import { defineMessages, FormattedMessage } from 'react-intl';
+import { withRouter } from 'react-router';
+
+import { getProgrammingFileURL } from 'lib/helpers/url-builders';
+
+import Editor from '../../../components/Editor';
+import ReadOnlyEditor from '../../../containers/ReadOnlyEditor';
+import { fileShape } from '../../../propTypes';
+
+const translations = defineMessages({
+  sizeTooBig: {
+    id: 'course.assessment.submission.answer.programming.sizeTooBig',
+    defaultMessage: 'The file is too big and cannot be displayed.',
+  },
+  downloadFile: {
+    id: 'course.assessment.submission.answer.programming.downloadFile',
+    defaultMessage: 'Download File',
+  },
+});
+
+const styles = {
+  warningIcon: {
+    display: 'inline-block',
+    verticalAlign: 'middle',
+  },
+};
+
+class ProgrammingFile extends React.Component {
+  renderProgrammingEditor() {
+    const { file, answerId, language } = this.props;
+    return (
+      <React.Fragment>
+        <h5>{file.filename}</h5>
+        <Editor
+          name={`${answerId}[content]`}
+          filename={file.filename}
+          language={language}
+        />
+      </React.Fragment>
+    );
+  }
+
+  renderReadOnlyProgrammingEditor() {
+    const { file, answerId } = this.props;
+    const { courseId, assessmentId, submissionId } = this.props.match.params;
+
+    const downloadLink = getProgrammingFileURL(courseId, assessmentId, submissionId,
+      parseInt(answerId.split('[')[0], 10), file.id);
+
+    if (file.content === null) {
```

Comment:
Just a quick check, if the programming file is empty (eg. empty template or an empty file was saved), will this be invoked?

Reply:
Empty file content will be empty string, not null.

Actually I used `content = nil` to represent the file being too big in the jbuilder, not sure if there's a better way to implement, maybe use an additional boolean field?